### ISSUE TITLE
Modify link existing node to bcast-link doesn't work

### ISSUE THREAD

**Issue by ibaldin** [RENCI-NRIG/orca5#76]
Creating a slice that has a standalone node and N nodes connected to a broadcast link and then as modify linking the node to that link does not do anything. Errors are not thrown, but the modify handler is not invoked and no new interface materializes on the reservation. Also pequod doesn't show any new modify properties on the VM reservation. Logs show what appears a normal sequence on the controller:

```
2016-10-12 20:24:13,422 [qtp821142925-35 - /orca/xmlrpc] INFO  controller.OrcaXmlrpcHandler - ORCA API modifySlice() invoked for slice-stitch-props
2016-10-12 20:24:13,422 [qtp821142925-35 - /orca/xmlrpc] INFO  controller.OrcaXmlrpcHandler - validateOrcaCredential(): Client presented a certificate with subject: EMAILADDRESS=ibaldin@email.unc.edu, CN=c7578309-b14c-4b4a-b555-1eff10f1b092
2016-10-12 20:24:13,424 [qtp821142925-35 - /orca/xmlrpc] INFO  controller.OrcaXmlrpcHandler - clientcertUserDN = [ibaldin@email.unc.edu, urn:publicid:IDN+ch.geni.net+user+ibaldin, urn:uuid:c7578309-b14c-4b4a-b555-1eff10f1b092]
2016-10-12 20:24:13,424 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - returning clientcertUserDN as userDN
2016-10-12 20:24:13,424 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.XmlrpcOrcaState - getSlice(): this slice ID=ead0f2ac-f427-4f8f-a505-c28551c95e2b for urn=slice-stitch-props
2016-10-12 20:24:13,455 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - discoverTypes: tamuvmsite/vlan rt=tamuvmsite.vlan available units=494
2016-10-12 20:24:13,455 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - Found abstract model for resource pool: tamuvmsite.vlan
2016-10-12 20:24:13,475 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - discoverTypes: tamuvmsite/vm rt=tamuvmsite.vm available units=41
2016-10-12 20:24:13,475 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - Found abstract model for resource pool: tamuvmsite.vm
2016-10-12 20:24:13,493 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - discoverTypes: tamuvmsite/lun rt=tamuvmsite.lun available units=58
2016-10-12 20:24:13,493 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - Found abstract model for resource pool: tamuvmsite.lun
2016-10-12 20:24:13,509 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - discoverTypes: tamuNet/vlan rt=tamuNet.vlan available units=10
2016-10-12 20:24:13,509 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - Found abstract model for resource pool: tamuNet.vlan
2016-10-12 20:24:13,993 [qtp821142925-35 - /orca/xmlrpc] INFO  controller.OrcaXmlrpcHandler - ORCA API modifySlice() started...
2016-10-12 20:24:14,045 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.XmlrpcOrcaState - getSlice(): this slice ID=ead0f2ac-f427-4f8f-a505-c28551c95e2b for urn=slice-stitch-props
2016-10-12 20:24:14,112 [qtp821142925-35 - /orca/xmlrpc] INFO  controller.OrcaXmlrpcHandler - modifySlice(): processing removed reservations
2016-10-12 20:24:14,113 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - No modifyremoved reservations in slice with urn slice-stitch-props sliceId  ead0f2ac-f427-4f8f-a505-c28551c95e2b;null;null
2016-10-12 20:24:14,113 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - No removed reservations in slice with urn slice-stitch-props sliceId  ead0f2ac-f427-4f8f-a505-c28551c95e2b
2016-10-12 20:24:14,113 [qtp821142925-35 - /orca/xmlrpc] INFO  controller.OrcaXmlrpcHandler - modifySlice(): processing add reservations
2016-10-12 20:24:14,113 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - No added reservations in slice with urn slice-stitch-props sliceId  ead0f2ac-f427-4f8f-a505-c28551c95e2b
2016-10-12 20:24:14,120 [qtp821142925-35 - /orca/xmlrpc] INFO  controller.SliceDeferThread - checkComputedReservations(): checkComputedReservaions slice-stitch-props/ead0f2ac-f427-4f8f-a505-c28551c95e2b has no delayed domains
2016-10-12 20:24:14,120 [qtp821142925-35 - /orca/xmlrpc] INFO  controller.SliceDeferThread - processSlice(): Processing slice slice-stitch-props/ead0f2ac-f427-4f8f-a505-c28551c95e2b immediately
2016-10-12 20:24:14,120 [qtp821142925-35 - /orca/xmlrpc] INFO  controller.SliceDeferThread - checkComputedReservations(): checkComputedReservaions slice-stitch-props/ead0f2ac-f427-4f8f-a505-c28551c95e2b has no delayed domains
2016-10-12 20:24:14,120 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.SliceDeferThread - demandSlice(): Issuing demand for reservation: 0b187f1c-3ff6-4afb-acea-d77455e176e1
2016-10-12 20:24:14,120 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.SliceDeferThread - demandSlice(): Issuing demand for reservation: 147d63d0-f89c-41f7-ad2d-a3c4328fe667
2016-10-12 20:24:14,120 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.SliceDeferThread - demandSlice(): Issuing demand for reservation: fad852e4-9056-4aa5-b822-6ff6068fcfad
2016-10-12 20:24:14,120 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.SliceDeferThread - demandSlice(): Issuing demand for reservation: b9e27b89-3fb4-48c1-9f79-e9ea03446304
2016-10-12 20:24:14,120 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.SliceDeferThread - demandSlice(): Issuing demand for reservation: c07e45ed-7ff4-4385-a507-7bc2b9efbb85
2016-10-12 20:24:14,120 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - modifySlice(): modifying existing reservations
2016-10-12 20:24:14,120 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - There are 1 modified reservations in the slice with urn slice-stitch-props sliceId = ead0f2ac-f427-4f8f-a505-c28551c95e2b
2016-10-12 20:24:14,120 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - Issuing modify demand for reservation: c07e45ed-7ff4-4385-a507-7bc2b9efbb85
2016-10-12 20:24:14,122 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - addActiveStatuWatch:numExistParent=1
2016-10-12 20:24:14,122 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - addActiveStatuWatch:[fad852e4-9056-4aa5-b822-6ff6068fcfad];;self=[c07e45ed-7ff4-4385-a507-7bc2b9efbb85]
2016-10-12 20:24:14,342 [qtp821142925-35 - /orca/xmlrpc] INFO  controller.OrcaXmlrpcHandler - Adding slice-stitch-props to modifiedSliceQ
2016-10-12 20:24:14,342 [qtp821142925-35 - /orca/xmlrpc] DEBUG controller.OrcaXmlrpcHandler - modifySlice(): returning result Here are the leases: 
Request id: ead0f2ac-f427-4f8f-a505-c28551c95e2b
[   Slice UID: ead0f2ac-f427-4f8f-a505-c28551c95e2b | Reservation UID: 0b187f1c-3ff6-4afb-acea-d77455e176e1 | Resource Type: tamuvmsite.vm | Resource Units: 1 ] 
[   Slice UID: ead0f2ac-f427-4f8f-a505-c28551c95e2b | Reservation UID: 147d63d0-f89c-41f7-ad2d-a3c4328fe667 | Resource Type: tamuvmsite.vm | Resource Units: 1 ] 
[   Slice UID: ead0f2ac-f427-4f8f-a505-c28551c95e2b | Reservation UID: fad852e4-9056-4aa5-b822-6ff6068fcfad | Resource Type: tamuvmsite.vlan | Resource Units: 1 ] 
[   Slice UID: ead0f2ac-f427-4f8f-a505-c28551c95e2b | Reservation UID: b9e27b89-3fb4-48c1-9f79-e9ea03446304 | Resource Type: tamuvmsite.vm | Resource Units: 1 ] 
[   Slice UID: ead0f2ac-f427-4f8f-a505-c28551c95e2b | Reservation UID: c07e45ed-7ff4-4385-a507-7bc2b9efbb85 | Resource Type: tamuvmsite.vm | Resource Units: 1 ] 
No errors reported
2016-10-12 20:24:15,320 [PeriodicPoolThread] INFO  controller.ReservationStatusUpdateThread - run(): Scanning active watch list
2016-10-12 20:24:15,346 [PeriodicPoolThread] DEBUG controller.ReservationStatusUpdateThread - run(): Removing active entries from watch 1
2016-10-12 20:24:15,346 [PeriodicPoolThread] INFO  controller.ReservationStatusUpdateThread - run(): processing 1 triggered active callbacks
2016-10-12 20:24:15,346 [PeriodicPoolThread] DEBUG controller.ReservationStatusUpdateThread - Invoking success callback for reservations [fad852e4-9056-4aa5-b822-6ff6068fcfad]

```
@YufengXin @paul-ruth @anriban

**Comment by ibaldin** [comment]
Nothing happens after the last log message above - i.e. the callback log is not visible.

**Comment by ibaldin** [comment]
Tested on TAMU and SM local controllers

**Comment by YufengXin** [comment]
Yeah, there appeared a problem in generating the correct properties for this case, which somehow didn’t count the existing vlan reservation correctly.
The fix is actually minor, the large chunk of “git diff” is due to adjusting the indent.

Ilya, the property “sitehostinterface” you asked is actually the sequence number of the NIC of VM.

Fixed in Commit #f9117a042b82ea9530ee8644dcaae76b2bf4acb5

-Yufeng



Yufeng Xin, PhD
RENCI
UNC at Chapel Hill
1-919-445-9633
yxin@renci.org <mailto:yxin@renci.org>

**Comment by ibaldin** [comment]
@YufengXin reports that when connecting a node to a net crossconnect that is part of a two-domain dumbbell he receives an error from Flukes. To be investigated.

**Comment by ibaldin** [comment]
More info about this bug:

![screen shot 2016-10-27 at 3 50 10 pm](https://cloud.githubusercontent.com/assets/12304033/19782729/184f6db2-9c5d-11e6-9ee4-8c99742097e8.png)

I suspect that the issue may be with multiple types associated with the corssconnect in question in the manifest.

**Comment by ibaldin** [comment]
Yes the issue is that the VLAN that is being connected to via link-add-modify to an existing node in the same domain has two type declarations:

```
  <rdf:Description rdf:about="http://geni-orca.renci.org/owl/bbnNet.rdf#bbnNet/Domain/vlan/9b8c5bc3-58eb-4e2d-a536-b0a4c6cef42c/vlan">
    <rdf:type rdf:resource="http://geni-orca.renci.org/owl/topology.owl#NetworkConnection"/>
    <rdf:type rdf:resource="http://geni-orca.renci.org/owl/topology.owl#CrossConnect"/>
```

while a regular VLAN has only one
```
  <rdf:Description rdf:about="http://geni-orca.renci.org/owl/fiuNet.rdf#fiuNet/Domain/vlan/902092b7-2ef7-42a6-be20-e1f1fc1211d4/vlan">
    <rdf:type rdf:resource="http://geni-orca.renci.org/owl/topology.owl#CrossConnect"/>
```
this confuses the parser.

**Comment by ibaldin** [comment]
Should be fixed in cba483a30057f3fec4f505f270b0d837cc9aae1d. This is NdlManifestParser, so affects Flukes and Ahab. I tried to be careful not to change existing behavior outside of the intended fix. New version of ndllib is deployed to Nexus. New version of flukes deployed to main webstart area (together with slice stitching fixes).

@paul-ruth
